File: incre-tests/synduce/constraints/all_positive/list_mps.f
Inductive CList = cnil Unit | single Int | concat {CList, CList};
Inductive List = nil Unit | cons {Int, List};

allpos = fix (
  \f: CList -> Bool. \c: CList.
  match c with
    cnil _ -> true
  | single w -> (> w 0)
  | concat {l, r} -> and (f l) (f r)
  end
);

cat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

repr = fix (
  \f: CList -> List. \cl: CList.
  match cl with
    cnil _ -> nil unit
  | single h -> cons {h, nil unit}
  | concat {l, r} -> cat (f l) (f r)
  end
);

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if < a b then b else a;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> max (+ (f t) h) 0
  end
);

target = fix (
  \f: CList -> Compress CList. \xs: CList.
  match xs with
    cnil _ -> cnil unit
  | single h -> single h
  | concat {l, r} -> concat {f l, f r}
  end
);

main = \xs: CList. if allpos xs then spec (repr (target xs)) else 0;

File: ml-benchmark/synduce/constraints/all_positive/list_mps.ml
type clist = Cnil of unit | Single of int | Concat of clist * clist
type llist = Nil of unit | Cons of int * llist

let rec allpos c =
  match c with
  | Cnil _ -> true
  | Single w -> w > 0
  | Concat (l, r) -> allpos l && allpos r

let rec cat x y =
  match x with
  | Cons (h, t) -> Cons (h, cat t y)
  | Nil _ -> y

let rec repr cl =
  match cl with
  | Cnil _ -> Nil ()
  | Single h -> Cons (h, Nil ())
  | Concat (l, r) -> cat (repr l) (repr r)

let min a b = if a < b then a else b
let max a b = if a < b then b else a

let rec spec xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) -> max ((spec t) + h) 0

val target: clist -> clist compress
let rec target xs =
  match xs with
  | Cnil _ -> Cnil ()
  | Single h -> Single h
  | Concat (l, r) -> Concat (target l, target r)

let program xs =
  if allpos xs then spec (repr (target xs)) else 0